M3-21 ListLinodes view.

rjstires opened the pull request:
## Purpose
Initial pass at ListLinodes.

## Notes
- Typed Linode, LinodeType, Image, TodoAny (moved out of `src/utils`).
- Typed the Redux state, left in mock data for testing purposes (Linodes, Types, Images).
- I honestly don't know why, but I'm using types that havent been imported/exported from `types.ts`
- Ignored that stupid Boolean rule.

## Airing of the Grievances
I'm not sure how I feel about this default props stuff...

asauber commented on src/types.ts:
```diff
@@ -5,3 +5,159 @@
+
+type TodoAny = any;
+
+/******************************************************************************
```
Rather than start a huge file with these "banners", I think we should have a src/types/ directory with e.g. src/types/linodes.ts

asauber commented on src/types.ts:
```diff
@@ -5,3 +5,159 @@
+  created: string;
+  updated: string;
+}
+type LinodeBackupDiskStatus = 'offline'
```
Nice detail on all these types!

asauber commented on src/App.tsx:
```diff
@@ -58,10 +63,9 @@ class App extends React.Component<Props> {
           <main className={classes.content}>
-            <Typography variant="display1" noWrap>
               <Switch>
                 <Route exact path="/dashboard" render={() => 'Dashboard'} />
```
Can we keep the `display1` style for these other routes using a small render function in this file?

asauber commented on src/features/linodes/ListLinodes/LinodeRow.tsx:
```diff
@@ -0,0 +1,162 @@
+};
+
+function clip(value: string): void {
+  console.log('clip', value);
```
stray console.log

asauber commented on src/features/linodes/ListLinodes/index.tsx:
```diff
@@ -0,0 +1,152 @@
+  },
+});
+
+interface Props extends StyledComponentProps<any> {
```
This could typing `WithStyles`

asauber commented on src/features/linodes/ListLinodes/LinodeRow.tsx:
```diff
@@ -0,0 +1,161 @@
+  return `${countryCode.toUpperCase()} ${titlecase(area)}`;
+}
+
+interface Props extends StyledComponentProps<'copyIcon' | 'inlineItems'> {
```
This could be `WithStyles`

asauber commented on src/features/linodes/ListLinodes/LinodeRow.tsx:
```diff
@@ -0,0 +1,161 @@
+  classes: {};
+}
+
+type PropsWithDefaults = Props & DefaultProps;
```
An intersection type isn't correct here, if we want to say that the props for this component are `Props` or they are the `DefaultProps`, then we should use a union type `Props | DefaultProps`. As soon as you do that, you will notice that you get type errors like:

```
Property 'memory' does not exist on type '{} | LinodeType'.
  Property 'memory' does not exist on type '{}'.
```

This is correct, with the default props, the component will crash. This indicates that the component needs to be modified to handle the default state appropriately.

asauber replied:
One way forward is to make the `Props` interface not have optional members. This is a presentational component rather than one that makes requests or holds state, so I think that this is a reasonable assumption.

rjstires commented:
I made some significant changes;
- Created a WithDocumentation component.
- Added some testing.
- Removed defaults from components.
- Corrected some typings.

asauber commented:
Thanks for removing those default props.

asauber commented:
When clicking the copy button for the IPv4 address, it copies the IPv6 address.

asauber commented on src/features/linodes/ListLinodes/ListLinodesEmptyState.tsx:
```diff
@@ -0,0 +1,67 @@
+import Typography from 'material-ui/Typography';
+import Button from 'material-ui/Button';
+
+const styles: StyleRulesCallback = (theme: Theme) => ({
```
Is there a reason not a parameterize this instance of `StyleRulesCallback`?

rjstires replied:
Im going to type it, but I havent really seen the "why" this needs to be typed. It doesn't check the style object, and I still get type completion in the implementation.

asauber commented on src/features/linodes/ListLinodes/index.tsx:
```diff
@@ -0,0 +1,87 @@
+import ListLinodesEmptyState from './ListLinodesEmptyState';
+
+const styles: StyleRulesCallback<'root'> = (theme: Theme) => ({
+  root: {
```
looks like this style isn't being used

rjstires replied:
Good eye, thanks.

asauber commented on src/types/index.ts:
```diff
@@ -0,0 +1,31 @@
+
+
+namespace Linode {
+  export type TodoAny = any;
```
I'd rather have this type outside any namespace

rjstires replied:
Any particular reason why?

asauber commented on src/features/linodes/ListLinodes/ListLinodesEmptyState.tsx:
```diff
@@ -0,0 +1,67 @@
+
+const styles: StyleRulesCallback = (theme: Theme) => ({
+  root: {
+    paddingTop: '25px',
```
these measurements should make use of our theme, e.g. `theme.spacing.unit * 3`. this way we can modify the theme and changes will apply throughout. see the first example in the docs https://material-ui-next.com/demos/tables/

asauber commented:
Looks great. After this round of changes, feel free to merge.
